HHS/simpler-grants-gov
# [Issue 2447] add relevancy sort option to opportunity search

## Summary
Fixes #2447

### Time to review: __20 mins__

## Changes proposed

Following note from the ticket at https://github.com/HHS/simpler-grants-gov/issues/2447#issuecomment-2427653490, this change:

* updates the default sort option on the opportunities search to "relevancy" (with a built in secondary sort of "post date descending", as previously implemented on the API)
* creates a new sort option in the UI titled "Default / Most Relevant" to be used as the default sort option

Also includes a couple of updates for stronger typing and improve testing of related components.

Note requirements update after review:

* update sort select option labels with new caps scheme
* add a "Sort by" label to the left of the sort select, shifting to above the select at mobile widths

## Context for reviewers


### Testing UI behavior

1. ensure you have local opensearch seeded (if not, run `make populate-search-opportunities` in the api directory)
1. visit local search page
2. _VERIFY_ the default selected sort option is "Default / Most Relevant"
2. _VERIFY_ search results are sorted in reverse chronological order by post date
3. enter a search term such as "grant" and click submit
4. _VERIFY_ search results are sorted by relevance in reverse chronological order by post date (since relevance is not visible directly in the UI, to verify further, see below)
5. select a new sort option
6. _VERIFY_ results are now sorted by your selected option
7. hard refresh or copy and paste the url for the page into a new incognito browser
8. _VERIFY_ sort selection is retained

### Testing sort results

1. add `relevancy_score = fields.Decimal()` on line 310 of `opportunity_schemas.py` and rebuild and restart the API container. `make stop && make build && make start` in the API directory should do it
9. put a debugger at line 66 of `SearchOpportunityApi.ts` and load the search page, performing various searches adn sorts
10. looking at the response you can confirm that the response array is sorted primarily on descending relevancy, and secondarily on descending post_date by default, but updates as you choose other options

## Additional information

### Screenshots
<img width="1223" alt="Screenshot 2024-10-22 at 1 17 28 PM" src="https://github.com/user-attachments/assets/d2588c9b-4f66-4e5b-bfea-c738f29c160d">

![Screenshot 2024-10-22 at 9 49 21 AM](https://github.com/user-attachments/assets/ed05f3a6-ddcd-42f5-a2ba-a5e6faa0c5f1)
